Universal property of Cauchy pseudocompletions and short maps of pseudometric spaces (#1818)
UniMath/SymmetryBookFormalization: src/metric-spaces/universal-property-short-maps-cauchy-pseudocompletions-of-pseudometric-spaces.lagda.md

Reviewer: all of the refactoring work you've done here should go in a separate PR to the addition of the universal property

> # Universal property of Cauchy pseudocompletions of pseudometric spaces and short maps into metric spaces

Reviewer 2: ```suggestion
# The universal property of Cauchy pseudocompletions of pseudometric spaces and short maps into metric spaces
```

open import metric-spaces.limits-of-cauchy-approximations-metric-spaces
open import metric-spaces.maps-pseudometric-spaces
open import metric-spaces.metric-spaces
open import metric-spaces.precomplete-short-maps-pseudometric-spaces
open import metric-spaces.pseudometric-spaces
open import metric-spaces.short-maps-metric-spaces
open import metric-spaces.short-maps-pseudometric-spaces
open import metric-spaces.similarity-of-elements-pseudometric-spaces

open import orthogonal-factorization-systems.extensions-maps
```

</details>

## Idea

Given a [metric space](metric-spaces.metric-spaces.md) `M` and a
[pseudometric space](metric-spaces.pseudometric-spaces.md) `P`, precomposition
with the unit of the
[Cauchy pseudocompletion](metric-spaces.cauchy-pseudocompletions-of-pseudometric-spaces.md)

```text
  κ : P → C P
```

maps [short maps](metric-spaces.short-maps-pseudometric-spaces.md) `g : C P → M`
to short maps `g ∘ κ : P → M`. For any
[Cauchy approximation](metric-spaces.cauchy-approximations-pseudometric-spaces.md)
`u : C P`, its
[image](metric-spaces.functoriality-short-maps-cauchy-pseudocompletions-of-pseudometric-spaces.md)

> `C(g ∘ κ) u : C M`

Reviewer 2: ```suggestion
`C (g ∘ κ) u : C M`
```

Conversely, if a short map `f : P → M`,
[extends](orthogonal-factorization-systems.extensions-maps.md) along `κ`, i.e.,
there exists a short map `g : C P → M` such that

```text
  f ~ g ∘ κ
```

then such an
{{#concept "extension" Disambiguation="of short maps on pseudometric spaces along κ" Agda=extension-short-map-cauchy-pseudocompletion-Pseudometric-Space}}
is unique and exists if and only if `f` is precomplete.

It follows that the type of
{{#concept "extensible short maps" Disambiguation="from a pseudometric space along κ" Agda=extensible-short-map-cauchy-pseudocompletion-Pseudometric-Space}}
is [equivalent](foundation.equivalences.md) to type of precomplete short maps.

Equivalently, the Cauchy pseudocompletion satisfies the
{{#concept "universal property" Disambiguation="of Cauchy pseudocompletions and precomplete short maps" Agda=is-contr-extension-precomplete-short-map-Pseudometric-Space}}
of Cauchy precompletions and precomplete short maps: for any precomplete short
map `f : P → M` from a pseudometric space in a metric space, there
[uniquely exists](foundation.uniqueness-quantification.md) an extension of `f`
along `κ`.

## Definitions

> ### The property of being the extension of a short map along the unit of Cauchy pseudocompletions

Reviewer 2: ```suggestion
### The property of being an extension of a short map along the unit of Cauchy pseudocompletions
```

        ( f))
      ( map-short-map-Pseudometric-Space
        ( cauchy-pseudocompletion-Pseudometric-Space P)
        ( pseudometric-Metric-Space M)
        ( g))

  is-prop-is-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space :
    is-prop
      is-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space
  is-prop-is-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space =
    is-prop-Π
      ( λ x →
        is-set-type-Metric-Space M
          ( map-short-map-Pseudometric-Space
            ( P)
            ( pseudometric-Metric-Space M)
            ( f)
            ( x))
          ( map-short-map-Pseudometric-Space
            ( cauchy-pseudocompletion-Pseudometric-Space P)
            ( pseudometric-Metric-Space M)
            ( g)
            ( map-unit-cauchy-pseudocompletion-Pseudometric-Space P x)))

  is-extension-prop-short-map-cauchy-pseudocompletion-Pseudometric-Space :
    Prop (l1 ⊔ l1')
  is-extension-prop-short-map-cauchy-pseudocompletion-Pseudometric-Space =
    ( is-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space ,
      is-prop-is-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space)
```

> ### Extensions of short maps along the unit map of Cauchy pseudocompletions

Reviewer 2: ```suggestion
### Extensions of short maps along the unit map of a Cauchy pseudocompletion
```

    short-map-Pseudometric-Space P (pseudometric-Metric-Space M)
  precomp-short-map-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space
    =
    precomp-short-map-unit-cauchy-pseudocompletion-Pseudometric-Space
      ( P)
      ( M)
      ( short-map-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space)

  map-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space :
    map-Pseudometric-Space
      ( cauchy-pseudocompletion-Pseudometric-Space P)
      ( pseudometric-Metric-Space M)
  map-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space =
    map-short-map-Pseudometric-Space
      ( cauchy-pseudocompletion-Pseudometric-Space P)
      ( pseudometric-Metric-Space M)
      ( short-map-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space)

  is-extension-map-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space :
    is-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space P M f
      ( short-map-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space)
  is-extension-map-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space
    = pr2 g
```

> ### Homotopies between of extensions of a short map

Reviewer 2: ```suggestion
### Homotopies between extensions of a short map
```

```agda
module _
  {l1 l2 l1' l2' : Level}
  (P : Pseudometric-Space l1 l2)
  (M : Metric-Space l1' l2')
  (f : short-map-Pseudometric-Space P (pseudometric-Metric-Space M))
  (g h : extension-short-map-cauchy-pseudocompletion-Pseudometric-Space P M f)
  where

  htpy-map-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space :
    UU (l1 ⊔ l2 ⊔ l1')
  htpy-map-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space =
    htpy-map-short-map-Pseudometric-Space
      ( cauchy-pseudocompletion-Pseudometric-Space P)
      ( pseudometric-Metric-Space M)
      ( short-map-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space
        ( P)
        ( M)
        ( f)
        ( g))
      ( short-map-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space
        ( P)
        ( M)
        ( f)
        ( h))
```

> ### Extensible short maps along the unit map of Cauchy pseudocompletions

Reviewer 2: ```suggestion
### Extendable short maps along the unit map of Cauchy pseudocompletions
```

Author: Should I change the names `[...]-extensible-short-map-[...]` to `[...]-extendable-short-map-[...]` too?

Reviewer 2: Ah, yes that might be good. "Extensible" and "Extendable" have slightly different meanings; "Extendable" means having an extension, which is what I think you meant here, right?

Author: > Ah, yes that might be good. "Extensible" and "Extendable" have slightly different meanings; "Extendable" means having an extension, which is what I think you meant here, right?

It's what I mean, yes. 

What's the other meaning of "extensible" in this context? (Sorry, you know English is not my native language 😅)

Reviewer 2: Extensible sounds more like software to me, something that is open-ended and amenable by future developers. English also isn't my first language, so I could be very off. I just don't recall seeing the word "extensible" in the context of maps that have extensions. I could even be very off on this one.

Author: my computer's dictionary defines _extensible_ as "able to be extended or stretched" and mentions _extendable_ as a synonym so I guess both would be ok. 

I personally like _extensible_ better, it resounds with _extension_ slightly better IMHO. 
If you really prefer _extendable_, I'll change the headers and names accordingly.

Reviewer 2: No that's ok, I can live with extensible:) Thanks for the extra effort!

```agda
module _
  {l1 l2 l1' l2' : Level}
  (P : Pseudometric-Space l1 l2)
  (M : Metric-Space l1' l2')
  where

  extensible-short-map-cauchy-pseudocompletion-Pseudometric-Space :
    UU (l1 ⊔ l2 ⊔ l1' ⊔ l2')
  extensible-short-map-cauchy-pseudocompletion-Pseudometric-Space =
    Σ ( short-map-Pseudometric-Space P (pseudometric-Metric-Space M))
      ( extension-short-map-cauchy-pseudocompletion-Pseudometric-Space P M)

module _
  {l1 l2 l1' l2' : Level}
  (P : Pseudometric-Space l1 l2)
  (M : Metric-Space l1' l2')
  (f : extensible-short-map-cauchy-pseudocompletion-Pseudometric-Space P M)
  where

  short-map-extensible-short-map-cauchy-pseudocompletion-Pseudometric-Space :
    short-map-Pseudometric-Space P (pseudometric-Metric-Space M)
  short-map-extensible-short-map-cauchy-pseudocompletion-Pseudometric-Space
    = pr1 f

>   exten-short-map-extensible-short-map-cauchy-pseudocompletion-Pseudometric-Space :

Reviewer 2: ```suggestion
  extension-short-map-extensible-short-map-cauchy-pseudocompletion-Pseudometric-Space :
```

Author: see https://github.com/UniMath/agda-unimath/pull/1818#discussion_r3171626600

        ( M)
        ( f)
        ( g)
        ( u) ,
      is-lim-map-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space
        ( P)
        ( M)
        ( f)
        ( g)
        ( u))

  exten-short-map-cauchy-pseudocompletion-is-precomplete-short-map-Pseudometric-Space :
    is-precomplete-short-map-Pseudometric-Space P M f →
    extension-short-map-cauchy-pseudocompletion-Pseudometric-Space P M f
  exten-short-map-cauchy-pseudocompletion-is-precomplete-short-map-Pseudometric-Space
    H =
    exten-precomplete-short-map-cauchy-pseudocompletion-Pseudometric-Space P M
      ( f , H)

  iff-is-precomplete-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space :
    extension-short-map-cauchy-pseudocompletion-Pseudometric-Space P M f ↔
    is-precomplete-short-map-Pseudometric-Space P M f
  pr1
    iff-is-precomplete-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space
    =
    is-precomplete-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space
  pr2
    iff-is-precomplete-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space
    =
    exten-short-map-cauchy-pseudocompletion-is-precomplete-short-map-Pseudometric-Space
```

> ### Extensible short maps are equivalent to precomplete short maps

Reviewer 2: ```suggestion
### Extendable short maps are equivalent to precomplete short maps
```

```agda
module _
  {l1 l2 l1' l2' : Level}
  (P : Pseudometric-Space l1 l2)
  (M : Metric-Space l1' l2')
  where

  equiv-precomplete-extensible-short-map-Pseudometric-Space :
    extensible-short-map-cauchy-pseudocompletion-Pseudometric-Space P M ≃
    precomplete-short-map-Pseudometric-Space P M
  equiv-precomplete-extensible-short-map-Pseudometric-Space =
    equiv-type-subtype
      ( is-prop-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space
        ( P)
        ( M))
      ( is-prop-is-precomplete-short-map-Pseudometric-Space P M)
      ( is-precomplete-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space
        ( P)
        ( M))
      ( exten-short-map-cauchy-pseudocompletion-is-precomplete-short-map-Pseudometric-Space
        ( P)
        ( M))
```

> ### Extensible short maps are the precompositions of their extensions

Reviewer 2: ```suggestion
### Extendable short maps are the precompositions of their extensions
```

```agda
module _
  {l1 l2 l1' l2' : Level}
  (P : Pseudometric-Space l1 l2)
  (M : Metric-Space l1' l2')
  where

  equiv-extensible-short-map-cauchy-pseudocompletion-Pseudometric-Space :
    extensible-short-map-cauchy-pseudocompletion-Pseudometric-Space P M ≃
    short-map-Pseudometric-Space
      ( cauchy-pseudocompletion-Pseudometric-Space P)
      ( pseudometric-Metric-Space M)
  equiv-extensible-short-map-cauchy-pseudocompletion-Pseudometric-Space =
    ( equiv-exten-precomplete-short-map-cauchy-pseudocompletion-Pseudometric-Space
      ( P)
      ( M)) ∘e
    ( equiv-precomplete-extensible-short-map-Pseudometric-Space P M)
```

> ### The type of extensions of a precomplete short map is contractible

Reviewer 2: Nice work!

      ( P)
      ( M)
      ( short-map-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space)

  map-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space :
    map-Pseudometric-Space
      ( cauchy-pseudocompletion-Pseudometric-Space P)
      ( pseudometric-Metric-Space M)
  map-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space =
    map-short-map-Pseudometric-Space
      ( cauchy-pseudocompletion-Pseudometric-Space P)
      ( pseudometric-Metric-Space M)
      ( short-map-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space)

  htpy-map-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space :
    map-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space ∘
    map-unit-cauchy-pseudocompletion-Pseudometric-Space P ~
    map-short-map-Pseudometric-Space P (pseudometric-Metric-Space M) f
  htpy-map-extension-short-map-cauchy-pseudocompletion-Pseudometric-Space =
    pr2 g
```

### Extensible short maps along the unit map of Cauchy pseudocompletions

```agda
module _
  {l1 l2 l1' l2' : Level}
  (P : Pseudometric-Space l1 l2)
  (M : Metric-Space l1' l2')
  where

>   extensible-short-map-cauchy-pseudocompletion-Pseudometric-Space :

Reviewer 2: Whenever you feel you've sufficiently addressed this comment, feel free to mark it as resolved.

      exten-short-map-extensible-short-map-cauchy-pseudocompletion-Pseudometric-Space
```

### The property of being a precomplete short map from a pseudometric space to a metric space

```agda
module _
  {l1 l2 l1' l2' : Level}
  (P : Pseudometric-Space l1 l2)
  (M : Metric-Space l1' l2')
  (f : short-map-Pseudometric-Space P (pseudometric-Metric-Space M))
  where

  is-precomplete-prop-short-map-Pseudometric-Space : Prop (l1 ⊔ l2 ⊔ l1' ⊔ l2')
  is-precomplete-prop-short-map-Pseudometric-Space =
    Π-Prop
      ( cauchy-approximation-Pseudometric-Space P)
      ( is-convergent-prop-cauchy-approximation-Metric-Space M ∘
        map-short-map-cauchy-pseudocompletion-Pseudometric-Space
          ( P)
          ( pseudometric-Metric-Space M)
          ( f))

  is-precomplete-short-map-Pseudometric-Space : UU (l1 ⊔ l2 ⊔ l1' ⊔ l2')
  is-precomplete-short-map-Pseudometric-Space =
    type-Prop is-precomplete-prop-short-map-Pseudometric-Space

  is-prop-is-precomplete-short-map-Pseudometric-Space :
    is-prop is-precomplete-short-map-Pseudometric-Space
  is-prop-is-precomplete-short-map-Pseudometric-Space =
    is-prop-type-Prop is-precomplete-prop-short-map-Pseudometric-Space
```

> ### The type of precomplete short maps from a pseudometric space to a metric space

Author: Or would you like me to split this PR and do `precomplete-short-maps-[...]` first, and come back here later for the "universal property of Cauchy pseudocompletions"?
